utils/format_doc.ml

```diff
@@ -0,0 +1,476 @@
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Florian Angeletti, projet Cambium, Inria Paris             *)
+(*                                                                        *)
+(*   Copyright 2021 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+
+type box_type =
+  | H
+  | V
+  | HV
+  | HoV
+  | B
+
+type stag = Format.stag
+
+type element =
+  | Text of string
+  | With_size of int
+  | Open_box of { kind: box_type ; indent:int }
+  | Close_box
+  | Open_tag of Format.stag
+  | Close_tag
+  | Open_tbox
+  | Tab_break of { width : int; offset : int }
+  | Set_tab
+  | Close_tbox
+  | Simple_break of { spaces : int; indent: int }
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
+
+  | Deprecated of (Format.formatter -> unit)
+
+type doc = { rev:element list } [@@unboxed]
+
+type t = doc
+
+let empty : doc = { rev = [] }
+
+let to_list doc = List.rev doc.rev
+let add doc x = { rev = x :: doc.rev }
+let fold f acc doc = List.fold_left f acc (to_list doc)
+let append left right = { rev = right.rev @ left.rev }
+
+let format_open_box_gen ppf kind indent =
+  match kind with
+  | H-> Format.pp_open_hbox ppf ()
+  | V -> Format.pp_open_vbox ppf indent
+  | HV -> Format.pp_open_hvbox ppf indent
+  | HoV -> Format.pp_open_hovbox ppf indent
+  | B -> Format.pp_open_box ppf indent
+
+let interpret_elt ppf = function
+  | Text x -> Format.pp_print_string ppf x
+  | Open_box { kind; indent } -> format_open_box_gen ppf kind indent
+  | Close_box -> Format.pp_close_box ppf ()
+  | Open_tag tag -> Format.pp_open_stag ppf tag
+  | Close_tag -> Format.pp_close_stag ppf ()
+  | Open_tbox -> Format.pp_open_tbox ppf ()
+  | Tab_break {width;offset} -> Format.pp_print_tbreak ppf width offset
+  | Set_tab -> Format.pp_set_tab ppf ()
+  | Close_tbox -> Format.pp_close_tbox ppf ()
+  | Simple_break {spaces;indent} -> Format.pp_print_break ppf spaces indent
+  | Break {fits;breaks} -> Format.pp_print_custom_break ppf ~fits ~breaks
+  | Flush {newline=true} -> Format.pp_print_newline ppf ()
+  | Flush {newline=false} -> Format.pp_print_flush ppf ()
+  | Newline -> Format.pp_force_newline ppf ()
+  | If_newline -> Format.pp_print_if_newline ppf ()
+  | With_size _ ->  ()
```

Comment:
Some functions like `pp_print_substring_as`, `with_size` and `if_newline` are not used. Is offering an alternative API to Format a goal ?
Otherwise, why not keep it smaller to help maintenance ?

You mentioned localization in https://github.com/ocaml/ocaml/pull/12182#issuecomment-2042180005. Could it be built on top of the this API ? Making it as small as needed could help adapting it later.

Reply:
My first goal was full source compatibility with `Format`, since this means compiler contributors don't need to know two APIs.

(And if we were to move away from `Format` compatibility, I would rather adopt `Fmt` naming convention)

For basic localization, a simple way to add support for compiler variant would be to add hooks for the format string interpreters since we only want to localize static texts.